Migration Request Integration Tests Concept

d-ylee opened the pull request:
For Issue ﻿#58 . @vkuznet This is the concept I came up with. Currently, I am using `cmsweb-testbed` for the migration origin.

Basically, the test goes as follows:

1. `bin/db_snapshot` starts the DBSMigrate and DBSMigration servers
    * Functionality for making a snapshot is commented out for now
2. Running `make test-migration-requests` starts the test for migration requests
3. Migration requests are stored in JSON files located in `MIGRATION_REQUEST_PATH` (currently `test/data/migration/requests`). These are parsed for the POST body for request to `/submit`
4. Migration request is submitted to local server
5. The status of the migration requests are requested via `/status`
6. Migration statuses are stored in array `requestStatus`. The index of the array corresponds to the migration status. The value is how many of the requests are in that status.
7. The status is checked every 2 seconds until no more migration requests are pending (0) or in progress (1)
8. The last step checks how many migration requests failed (index of 9). If this is greater than 0, then `t.Fatal` is thrown
    * I will probably also need to check the number of successful requests, as well as the other statuses.

What are your thoughts on this approach?

vkuznet commented:
@d-ylee , I think the logic is correct but I doubt it will work reliably, and it will not work in GitHub Actions. Here is my arguments:
- if we request migration request from cmsweb or cmsweb-testbed then it means that the node/test should have proper proxy certificate or token (later we will switch to token authentication). This will require to store in GitHub Action such credentials and it is security flow
- it we will use random datasets or blocks which we do not control we may end-up with very large dataset or block and the time for migration will grow which will lead that the test will stale
- the new logic of DBSMigrate server is asynchronous, i.e. it accepts requests, put them in queue, and DBSMigration server will pick them up (which has some polling cycle). As such the waiting time to complete the tests we will not be able to control

Said that, I think we should re-organize the test to have the following:
- setup stand-alone DBS server (which will serve the purpose of cmsweb or cmsweb-testbed). This server does not require any credentials as it will be run together (on the same localhost) as DBSMigrate/DBSMigration
   - this server should have its own DB back-end (i.e. its own SQLite DB)
- inject into this server datasets/blocks which we want to migrate. The migration should be done via `bulkblocks` API.
   - for that we will need to prepare json files which we will use with `bulkblocks` API. We can get any them independently from real DBSReader via `blockdump` API
- start DBSMigrate and DBSMigration servers with their own DB back-end (SQLiteDB)
- perform all steps from your proposal but use our own DBS server (which we can access from localhost)

Doing this way we will be independent from cmsweb or cmsweb-testebed, therefore no credentials will be required. We only need to prepare JSON files up-front (and can decide which one to use, i.e. we can only use small datasets or blocks but with similar complexity as real ones). These files we can prepare independently and keep them in repository. Therefore, such test can be performed in GitHub Actions as part of our test suit.

d-ylee commented:
@vkuznet Would this mean we will commit the SQLite DB into GitHub?

vkuznet commented:
@d-ylee , due to new async nature of migration server our current unit test fails with the following message:
```
2022/08/26 16:11:41 migrate_test.go:290: Received data [{"migration_details":{"migration_request_id":2,"migration_url":"https://cmsweb.cern.ch:8443/dbs/prod/global/DBSReader","migration_input":"/ZMM/Summer11-DESIGN42_V11_428_SLHC1-v1/GEN-SIM#8420a3e7-96cf-48b1-a6a8-f9abcd3de8ef","migration_status":5,"migration_server":"","create_by":"DBS-workflow","creation_date":1661544700,"last_modified_by":"DBS-workflow","last_modification_date":1661544700,"retry_count":0},"migration_report":"Migration request is started","status":"QUEUED","error":null}]
    migrate_test.go:295: invalid return status of migration request {MigrationRequest:{MIGRATION_REQUEST_ID:2 MIGRATION_URL:https://cmsweb.cern.ch:8443/dbs/prod/global/DBSReader MIGRATION_INPUT:/ZMM/Summer11-DESIGN42_V11_428_SLHC1-v1/GEN-SIM#8420a3e7-96cf-48b1-a6a8-f9abcd3de8ef MIGRATION_STATUS:5 MIGRATION_SERVER: CREATE_BY:DBS-workflow CREATION_DATE:1661544700 LAST_MODIFIED_BY:DBS-workflow LAST_MODIFICATION_DATE:1661544700 RETRY_COUNT:0} Report:Migration request is started Status:QUEUED Error:<nil>}
2022/08/26 16:11:41 migrate.go:721: Migration request {2 https://cmsweb.cern.ch:8443/dbs/prod/global/DBSReader /ZMM/Summer11-DESIGN42_V11_428_SLHC1-v1/GEN-SIM#8420a3e7-96cf-48b1-a6a8-f9abcd3de8ef 5  DBS-workflow 1661544700 DBS-workflow 1661544700 0} with context is cancelled context deadline exceeded
```
I introduced new `QUEUED` status which means that request is accepted. This is non-blocking response to the client, but we need to adjust code to wait until migration server will pick it up and process it. Then, we need to check final status of this test to check that it successfully processed. Please include this into your concept.

d-ylee commented:
@vkuznet I changed the migration requests integration tests to be locally run. It can be found under `test/int_migration_test.go`. The servers are run through the `bin/start_test_migration` script.

The initial data is defined in the `*_grandparent_block.json`, `*_parent_block.json`, and `*_block.json` in the `data/migration/requests` directory. I did blockdumps from the cmsweb production server, then removed any unused files and block data.

Due to sqlite3 database locking up if I check the migration status every 2 seconds, I extended the sleep time to 10 seconds.

vkuznet commented:
Dennis, does your PR is up to date with master? Before making a merge I wanted to test it and I did the following:
```
git status
On branch master
Your branch is up to date with 'origin/master'.

# download PR
curl -ksLO https://github.com/dmwm/dbs2go/pull/75.patch

# apply patch in dry-run mode
patch -p1 --dry-run < 75.patch
patching file .gitignore
patching file Makefile
patching file bin/db_snapshot
patching file test/data/migration/requests/README.md
patching file test/data/migration/requests/genericttbar.json
patching file test/migrate_test.go
Hunk #2 succeeded at 387 (offset 8 lines).
patching file .gitignore
patching file Makefile
patching file bin/db_snapshot
patching file test/data/migration/requests/README.md
patching file test/data/migration/requests/genericttbar.json
patching file test/migrate_test.go
Hunk #2 succeeded at 387 (offset 8 lines).
patching file .gitignore
patching file Makefile
Hunk #2 FAILED at 176.
1 out of 2 hunks FAILED -- saving rejects to file Makefile.rej
can't find file to patch at input line 621
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff --git a/bin/db_snapshot b/bin/db_snapshot
|index 1a48607..7b1cd19 100755
|--- a/bin/db_snapshot
|+++ b/bin/db_snapshot
--------------------------
File to patch:
```
So, it seems to me that your PR is not yet align with master branch as it produce rejected Makefile and unable to create db_snapshot. Could you please double check and let me know (you may test this procedure if you'll clone master) when it is ready.

d-ylee commented:
@vkuznet I think I fixed the issue. Please try it again.

vkuznet commented:
@d-ylee I can apply the patch now via `git apply 75.patch` command, but it does not work out of the box. Here the issues I foudn:
- the Makefile should not have commented lines, i.e. you must remove these lines:
```
	# rm -f /tmp/dbs-one.db && \
	# sqlite3 /tmp/dbs-one.db < ../static/schema/sqlite-schema.sql && \
	# echo sqlite3 /tmp/dbs-one.db sqlite > ./dbfile_1 && \
	# rm -f /tmp/dbs-two.db && \
	# sqlite3 /tmp/dbs-two.db < ../static/schema/sqlite-schema.sql && \
	# echo sqlite3 /tmp/dbs-two.db sqlite > ./dbfile_2 && \
```
otherwise it fails like this on my mac:
```
make test-migration-requests
...
	go test -v -failfast -timeout 60m -run MigrationRequests
/bin/sh: -c: line 1: syntax error: unexpected end of file
make: *** [test-migration-requests] Error 2
```
- Then, the `bin/start_test_migration` fails to start since it does not inherit the shell environment and in order to compile on Mac we need to have proper libraries for ORACLE. So I was able to patch it like this:
```
+#
+export DYLD_LIBRARY_PATH=/opt/Oracle/instantclient_19_8
+export PKG_CONFIG_PATH=$PWD
```
which points to my oracle instance client. The `DYLD_LIBRARY_PATH` is what mac uses instead of `LD_LIBRARY_PATH` on Linux
- after these changes I can run test which compiles the code but it fails to execute it:
```
make test-migration-requests
cd test && \
	cd .. && \
	./bin/start_test_migration && \
	cd test && \
	LD_LIBRARY_PATH=`cat /Users/vk/Work/Languages/Go/gopath/src/github.com/dmwm/dbs2go/oci8.pc | grep "libdir=" | sed -e "s,libdir=,,"` DYLD_LIBRARY_PATH=`cat /Users/vk/Work/Languages/Go/gopath/src/github.com/dmwm/dbs2go/oci8.pc | grep "libdir=" | sed -e "s,libdir=,,"` \
	DBS_API_PARAMETERS_FILE=../static/parameters.json \
	DBS_READER_LEXICON_FILE=../static/lexicon_reader.json \
	DBS_WRITER_LEXICON_FILE=../static/lexicon_writer.json \
	BULKBLOCKS_DATA_FILE=./data/migration/bulkblocks_data.json \
	MIGRATION_BLOCKS_DATA_FILE=./data/migration/requests/genericttbar_blockdump.json \
	MIGRATION_REQUESTS_PATH=./data/migration/requests \
	go test -v -failfast -timeout 60m -run MigrationRequests
building dbs2go
go clean; rm -rf pkg dbs2go*; go build -ldflags="-s -w -X main.gitVersion=`git describe --tags`"
/Users/vk/Work/Languages/Go/gopath/src/github.com/dmwm/dbs2go
=== RUN   TestMigrationRequests2
=== RUN   TestMigrationRequests2/Insert_initial_data
=== RUN   TestMigrationRequests2/Insert_initial_data/Grandparent_blocks
    int_migration_test.go:70: Post "http://localhost:8990/dbs-one-writer/bulkblocks": dial tcp [::1]:8990: connect: connection refused
--- FAIL: TestMigrationRequests2 (0.08s)
    --- FAIL: TestMigrationRequests2/Insert_initial_data (0.00s)
        --- FAIL: TestMigrationRequests2/Insert_initial_data/Grandparent_blocks (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x4673261]

goroutine 6 [running]:
testing.tRunner.func1.2({0x48431e0, 0x4e2ca50})
	/opt/local/lib/go/src/testing/testing.go:1396 +0x24e
testing.tRunner.func1()
	/opt/local/lib/go/src/testing/testing.go:1399 +0x39f
panic({0x48431e0, 0x4e2ca50})
	/opt/local/lib/go/src/runtime/panic.go:884 +0x212
github.com/dmwm/dbs2go/test.TestMigrationRequests2.func2.1(0xc0004511e0)
	/Users/vk/Work/Languages/Go/gopath/src/github.com/dmwm/dbs2go/test/int_migration_test.go:73 +0x1a1
testing.tRunner(0xc0004511e0, 0xc000424c40)
	/opt/local/lib/go/src/testing/testing.go:1446 +0x10b
created by testing.(*T).Run
	/opt/local/lib/go/src/testing/testing.go:1493 +0x35f
exit status 2
FAIL	github.com/dmwm/dbs2go/test	0.625s
make: *** [test-migration-requests] Error 1
```
From the error I see that it fails to start a server and talk to it: `dial tcp [::1]:8990: connect: connection refused`. Further, if I look at `/tmp/dbs2go_reader_1-vkair.log-20220915` log file I see the following:
```
[2022-09-15 16:46:53.441645 -0400 EDT m=+0.123538826] Configuration: {"port":8989,"staticdir":"./static","base":"/dbs-one-reader","verbose":0,"log_file":"/tmp/dbs2go_reader_1.log","utc":false,"monit_type":"","monit_producer":"","hmac":"","limiter_rate":"100-S","limiter_header":"","limiter_skip_list":null,"metrics_prefix":"dbs2go","server_type":"DBSReader","etag":"","cache_control":"","cms_role":null,"cms_group":null,"migration_dbfile":"","migration_server_interval":60,"migration_process_timeout":300,"migration_cleanup_interval":600,"migration_cleanup_offset":7776000,"migration_retries":3,"migration_async_timeout":600,"dbfile":"./test/dbfile_1","max_db_connections":1000,"max_idle_connections":100,"db_monitoring_interval":0,"api_parameters_file":"./static/parameters.json","lexicon_file":"./static/lexicon_reader.json","file_chunk_size":10,"file_lumi_chunk_size":500,"file_lumi_max_size":10000,"file_lumi_insert_method":"chunks","concurrent_bulkblocks":false,"templates":"./static/templates","jscripts":"","images":"","styles":"","serverkey":"","servercrt":"","rootCA":"","csrfKey":"","production":false,"tlsRefreshInterval":14400,"graphqlSchema":""}
[2022-09-15 16:46:53.444447 -0400 EDT m=+0.126340592] limiter rate='100-S'
[2022-09-15 16:46:53.444578 -0400 EDT m=+0.126471989] parse Config.DBFile: ./test/dbfile_1
[2022-09-15 16:46:53.444946 -0400 EDT m=+0.126840171] unable to open "sqlite3, error /tmp/dbs-one.db
[2022-09-15 16:46:53.445029 -0400 EDT m=+0.126922600] sql: unknown driver "\"sqlite3" (forgotten import?)
```
which leads me to conclude that there is an error with sqlite3 import. Finally, when I look at your logic in `bin/start_test_migration` I found that you create `dbfile_1/2` in current project area while it should creates them in test area. So you need to replace this line:
```
echo sqlite3 /tmp/dbs-one.db sqlite > ./dbfile_1
```
with this one:
```
echo sqlite3 /tmp/dbs-one.db sqlite > ./test/dbfile_1
```
and the same apply to `dbfile_2`.

After these changes I was able to run `make test-migration-requests` and got the desired outcome:
```
make test-migration-requests
...
=== RUN   TestMigrationRequests2/Test_migration/Check_Migration_Request_status
Number of Requests: 1, Successful: 0, Statuses: [3 0 0 0 0 0 0 0 0 0]
Number of Requests: 1, Successful: 0, Statuses: [3 0 0 0 0 0 0 0 0 0]
Number of Requests: 1, Successful: 0, Statuses: [3 0 0 0 0 0 0 0 0 0]
Number of Requests: 1, Successful: 0, Statuses: [3 0 0 0 0 0 0 0 0 0]
Number of Requests: 1, Successful: 0, Statuses: [3 0 0 0 0 0 0 0 0 0]
Number of Requests: 1, Successful: 2, Statuses: [0 1 2 0 0 0 0 0 0 0]
Number of Requests: 1, Successful: 2, Statuses: [0 1 2 0 0 0 0 0 0 0]
Number of Requests: 1, Successful: 2, Statuses: [0 1 2 0 0 0 0 0 0 0]
Number of Requests: 1, Successful: 2, Statuses: [0 1 2 0 0 0 0 0 0 0]
Number of Requests: 1, Successful: 2, Statuses: [0 1 2 0 0 0 0 0 0 0]
Number of Requests: 1, Successful: 2, Statuses: [0 1 2 0 0 0 0 0 0 0]
Number of Requests: 1, Successful: 3, Statuses: [0 0 3 0 0 0 0 0 0 0]
--- PASS: TestMigrationRequests2 (180.25s)
    --- PASS: TestMigrationRequests2/Insert_initial_data (0.09s)
        --- PASS: TestMigrationRequests2/Insert_initial_data/Grandparent_blocks (0.05s)
        --- PASS: TestMigrationRequests2/Insert_initial_data/Parent_blocks (0.02s)
        --- PASS: TestMigrationRequests2/Insert_initial_data/Blocks (0.02s)
    --- PASS: TestMigrationRequests2/Test_migration (120.10s)
        --- PASS: TestMigrationRequests2/Test_migration/Submit_Migration_Request (0.02s)
        --- PASS: TestMigrationRequests2/Test_migration/Process_migration_request,_id_=_1 (0.01s)
        --- PASS: TestMigrationRequests2/Test_migration/Check_Migration_Request_status (120.07s)
PASS
ok  	github.com/dmwm/dbs2go/test	180.618s
```
But I do not understand the output which says:
```
Number of Requests: 1, Successful: 3, Statuses: [0 0 3 0 0 0 0 0 0 0]
```
I think the number 3 at position 3 tells me that 3 migration requests finished successfully. But I do not understand whey this array has 10 entries. I think they corresponds to migration codes, right? If so can you change print statement to make it more clear.

Anyway, I hope you can adjust this PR based on my findings and I can try it again.

d-ylee commented:
@vkuznet For some reason, after I made modifications to comments, the action fails. This happens in the IntegrationTest of `Test primarydataset/primarydatasets GET with invalid parameter key`. Do validators get modified in the Github action?

vkuznet commented:
No, it is due to my recent change to dbs.DBSError struct where I introduced stacktrace, see https://github.com/dmwm/dbs2go/blob/master/dbs/errors.go#L76

I was on a meeting and didn't update test code to handle new change in dbs.DBSError struct. Now build is fine, please rebase.

d-ylee commented:
@vkuznet Does this look okay?

vkuznet commented:
Dennis, here is what I got when I applied your patch:
```
make test-migration-requests
LD_LIBRARY_PATH=`cat /Users/vk/Work/Languages/Go/gopath/src/github.com/dmwm/dbs2go/oci8.pc | grep "libdir=" | sed -e "s,libdir=,,"` DYLD_LIBRARY_PATH=`cat /Users/vk/Work/Languages/Go/gopath/src/github.com/dmwm/dbs2go/oci8.pc | grep "libdir=" | sed -e "s,libdir=,,"` \
	./bin/start_test_migration && \
	cd test && \
	LD_LIBRARY_PATH=`cat /Users/vk/Work/Languages/Go/gopath/src/github.com/dmwm/dbs2go/oci8.pc | grep "libdir=" | sed -e "s,libdir=,,"` DYLD_LIBRARY_PATH=`cat /Users/vk/Work/Languages/Go/gopath/src/github.com/dmwm/dbs2go/oci8.pc | grep "libdir=" | sed -e "s,libdir=,,"` \
	DBS_API_PARAMETERS_FILE=../static/parameters.json \
	DBS_READER_LEXICON_FILE=../static/lexicon_reader.json \
	DBS_WRITER_LEXICON_FILE=../static/lexicon_writer.json \
	BULKBLOCKS_DATA_FILE=./data/migration/bulkblocks_data.json \
	MIGRATION_BLOCKS_DATA_FILE=./data/migration/requests/genericttbar_blockdump.json \
	MIGRATION_REQUESTS_PATH=./data/migration/requests \
	go test -v -failfast -timeout 10m -run MigrationRequests

/opt/Oracle/instantclient_19_8
Building dbs2go...
go clean; rm -rf pkg dbs2go*; go build -ldflags="-s -w -X main.gitVersion=`git describe --tags`"
Starting dbs-one-reader
Starting dbs-one-writer
Starting dbs-two-reader
Starting dbs-two-writer
Starting dbs-migrate
Starting dbs-migration
=== RUN   TestMigrationRequests2
=== RUN   TestMigrationRequests2/Insert_initial_data
=== RUN   TestMigrationRequests2/Insert_initial_data/Grandparent_blocks
=== RUN   TestMigrationRequests2/Insert_initial_data/Parent_blocks
=== RUN   TestMigrationRequests2/Insert_initial_data/Blocks
=== RUN   TestMigrationRequests2/Test_migration
=== RUN   TestMigrationRequests2/Test_migration/Submit_Migration_Request
    int_migration_test.go:149: Migration Report: [{{1 http://localhost:8989/dbs-one-reader /GenericTTbar/HC-CMSSW_9_2_6_91X_mcRun1_realistic_v2-v2/AODSIM#3517e1b6-76e3-11e7-a0c8-02163e00d7b3 5  DBS-workflow 1663787168 DBS-workflow 1663787168 0} Migration request is started QUEUED <nil>}]
=== RUN   TestMigrationRequests2/Test_migration/Process_migration_request,_id_=_1
=== RUN   TestMigrationRequests2/Test_migration/Check_Migration_Request_status
Requests: 3, Successful: 0, Statuses: [3 0 0 0 0 0 0 0 0 0]
Requests: 3, Successful: 0, Statuses: [3 0 0 0 0 0 0 0 0 0]
Requests: 3, Successful: 0, Statuses: [3 0 0 0 0 0 0 0 0 0]
Requests: 3, Successful: 0, Statuses: [3 0 0 0 0 0 0 0 0 0]
Requests: 3, Successful: 0, Statuses: [3 0 0 0 0 0 0 0 0 0]
Requests: 3, Successful: 2, Statuses: [0 1 2 0 0 0 0 0 0 0]
Requests: 3, Successful: 2, Statuses: [0 1 2 0 0 0 0 0 0 0]
Requests: 3, Successful: 2, Statuses: [0 1 2 0 0 0 0 0 0 0]
Requests: 3, Successful: 2, Statuses: [0 1 2 0 0 0 0 0 0 0]
Requests: 3, Successful: 2, Statuses: [0 1 2 0 0 0 0 0 0 0]
Requests: 3, Successful: 2, Statuses: [0 1 2 0 0 0 0 0 0 0]
Requests: 3, Successful: 3, Statuses: [0 0 3 0 0 0 0 0 0 0]
--- PASS: TestMigrationRequests2 (120.36s)
    --- PASS: TestMigrationRequests2/Insert_initial_data (0.10s)
        --- PASS: TestMigrationRequests2/Insert_initial_data/Grandparent_blocks (0.05s)
        --- PASS: TestMigrationRequests2/Insert_initial_data/Parent_blocks (0.02s)
        --- PASS: TestMigrationRequests2/Insert_initial_data/Blocks (0.02s)
    --- PASS: TestMigrationRequests2/Test_migration (120.16s)
        --- PASS: TestMigrationRequests2/Test_migration/Submit_Migration_Request (0.02s)
        --- PASS: TestMigrationRequests2/Test_migration/Process_migration_request,_id_=_1 (0.01s)
        --- PASS: TestMigrationRequests2/Test_migration/Check_Migration_Request_status (120.13s)
=== RUN   TestMigrationRequests
=== RUN   TestMigrationRequests/Insert_initial_data
=== RUN   TestMigrationRequests/Insert_initial_data/Grandparent_blocks
    migration_test.go:717: Post "http://localhost:8990/dbs-one-writer/bulkblocks": dial tcp [::1]:8990: connect: connection refused
--- FAIL: TestMigrationRequests (0.11s)
    --- FAIL: TestMigrationRequests/Insert_initial_data (0.02s)
        --- FAIL: TestMigrationRequests/Insert_initial_data/Grandparent_blocks (0.02s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x10 pc=0x4683d81]

goroutine 82 [running]:
testing.tRunner.func1.2({0x48490a0, 0x4e34b30})
	/opt/local/lib/go/src/testing/testing.go:1396 +0x24e
testing.tRunner.func1()
	/opt/local/lib/go/src/testing/testing.go:1399 +0x39f
panic({0x48490a0, 0x4e34b30})
	/opt/local/lib/go/src/runtime/panic.go:884 +0x212
github.com/dmwm/dbs2go/test.TestMigrationRequests.func2.1(0xc0005029c0)
	/Users/vk/Work/Languages/Go/gopath/src/github.com/dmwm/dbs2go/test/migration_test.go:720 +0x1a1
testing.tRunner(0xc0005029c0, 0xc000060fc0)
	/opt/local/lib/go/src/testing/testing.go:1446 +0x10b
created by testing.(*T).Run
	/opt/local/lib/go/src/testing/testing.go:1493 +0x35f
exit status 2
FAIL	github.com/dmwm/dbs2go/test	120.921s
make: *** [test-migration-requests] Error 1
```

And, apart from that I also tweak your `bin/start_test_migration` to explicitly set DYLD_LIBRARY_PATH since on macOS it is not propagated from parent process, see https://stackoverflow.com/questions/60126159/how-to-set-ld-library-path-dyld-library-path-on-macos But this will be separate issue to address how to properly set this env, I think we should add something to Makefile to set it up via using `setup.sh` script. So far I can live with modification I made manually. But you should address the failure in the test.

d-ylee commented:
Hi Valentin, I actually do not have a `TestMigrationRequests2` anymore. I moved and renamed that file, which may still exist after you patched the new code.

I'll take a look into the DYLD_LIBRARY_PATH. I have an M1 mac, and Oracle did not provide the instantclient for Apple Silicone. I have been developing in a Linux Machine.

Are you tweaks the same as before where you put the path of the library here?
https://github.com/d-ylee/dbs2go/blob/aff2a1fa587ddf16849711546af9fc7515823d1f/bin/start_test_migration#L6-L7

I tried to use the `odir` variable in the `Makefile` to define the `*_LIBRARY_PATH`. Does this work on macOS?

vkuznet commented:
@d-ylee , I'm not sure how it is possible but if you look at https://patch-diff.githubusercontent.com/raw/dmwm/dbs2go/pull/75.patch then you can clearly see that `TestMigrationRequests2` is in a patch. Did you rebase your PR? Please double check that patch actually reflects the changes you have. I can't test it otherwise since this test is present in patch file.

vkuznet commented:
@d-ylee , I found your `test/int_migration_test.go` file in this commit https://github.com/dmwm/dbs2go/commit/89b3255bcf8be8bc01acf7ed1b83dbdd9c5b0d4f and it provides `TestMigrationRequests2` function. So you need to either discard this commit in your PR or do something with the logic of this test function.

vkuznet commented:
Please paste the exact steps you did. I'm applying patch using this command: `git apply 75.patch`

vkuznet commented:
ok, now we understand the issue I was facing. It is something in `git apply` rather in PR itself. I'll apply the patch and report later how it goes and it will run fine I'll merge it.

vkuznet commented on bin/start_test_migration:
```diff
@@ -1,45 +1,55 @@
-pwd
-
-# Start dbs-one-reader
+echo $LD_LIBRARY_PATH
```
Dennis, please remove line 3 to line 11 (the build section) which should not be part of the unit test suit. This will automatically resolve issue with DYLD, LD paths. We can locally and manually build dbs2go executable while start test migration script will only use it

vkuznet commented:
Dennis, I was able now to apply your patch and run it, It finished ok. I only requested to remove build part from start migration server script which will solve issue with DYLD,LD paths  in local setup When we need a `dbs2go` in GitHub Action, then we'll add appropriate section in build workflow file (which I think already have it).

d-ylee commented:
@vkuznet After I removed it, the script cannot find the `dbs2go` binary. Do I need to create a build job or step in the `go.yml` workflow?

vkuznet commented:
Yes, you need to add an additional step to go.yaml to make build, see examples in `build.yml` https://github.com/dmwm/dbs2go/blob/master/.github/workflows/build.yml#L23 Basically you'll two two section one is build and another is Test. The former will build the executable with make, and later will do the tests. In `go.yaml` file you can add PATH or any other environments to navigate through your setup.

d-ylee commented:
@vkuznet I think I got it working. Can you please check?

vkuznet commented on test/migrate_test.go:
```diff
@@ -384,3 +387,119 @@ func TestMigrateRemove(t *testing.T) {
+				for _, m := range migrationStatus {
+					requestStatus[m.MigrationStatus] += 1
+				}
+				fmt.Printf("Number of Requests: %d, Successful: %d, Statuses: %v\n", len(d), requestStatus[2], requestStatus)
```
Please add explanation about list of statuses, i.e. instead of `[0 0 3 0 0 0 0 0 0 0]` we should have something explaining what element of this list means. For example, 
```
Statuses:
QUEUED: 0
IN PROGRESS: 0
SUCCESS: 3
FAILED: 0
...
```

vkuznet commented on test/migration_test.go:
```diff
@@ -274,6 +280,49 @@ func TestIntMigration(t *testing.T) {
+			for _, m := range migrationStatus {
+				requestStatus[m.MigrationStatus] += 1
+			}
+			fmt.Printf("Requests: %d, Successful: %d, Statuses: %v\n", len(migrationStatus), requestStatus[2], requestStatus)
```
same argument here, we need explanation for Statuses array `[0 0 3 0 0 0 0 0 0 0]`

vkuznet reviewed (changes requested):
Dennis, this is very good and we almost done, but I want you to do two more actions. As I mentioned in a code we need explanation of Statuses. And, I want you to break down into individual tests the `test-github`, see https://github.com/dmwm/dbs2go/blob/41d7e8274174c8790fff0d331827c9a92a022a28/.github/workflows/go.yml#L43 The reason is that it becomes very cumbersome to read GitHub Action log which is huge when we run all tests. Instead, break down `test-github` section into individual sections, like:
- Test SQL
- Test Reader
- Test Writer
- Test Migration
And put into this section test relevant for them. This will allow us to view GtiHub Action log quickly to identify which test will fail.

d-ylee commented:
@vkuznet I separated the github tests into separate steps in the `go.yml` `build` job.

I have a question regarding the migration request test. The test in `test/migrate_test.go` is the original, `cmsweb-testbed` based migration request test, so it is not a local test.

https://github.com/d-ylee/dbs2go/blob/e27c572ff7419d810f775b3ca92bd2aeabdbc82d/test/migrate_test.go#L411

I spelled it incorrectly so that `make test-migration-requests` does not run it. We discussed that we wanted a local test for migration requests. Should we keep this, or can I remove it?

vkuznet commented:
Dennis, thanks for separation of tests, I think we will benefit from this separation in a long run.

Now about cmsweb, we should remove it from all tests. The problem with that is that we require to upload to github a secret with valid x509, moreover we're moving away from x509 authentication and will need tokens, etc. Bottom line, everything we're testing should be local. As I explained we only need some data which we may retrieve manually from DBS and store in our test data area (as we already do). And, then with data we should perform only local tests. Therefore, you should remove the cmsweb test now.
